README.md
# filterjson

a small service to remove|retain top level keys from a JSON structure


## usage

API is `POST /` accepting a body of `application/json` and query params `remove` and/or `filter`

* `filter` will keep only the field name given in `filter` (1..*)
* `remove` will remove the field name given in `remove` (1..*)

Accepts either a JSON object or an array of JSON objects (but not, say, an array 
of arrays)

examples with `remove`

```
curl -v -X POST -d @samples/array.json -H "content-type: application/json" 'http://localhost:12001?remove=name'
```

```
curl -v -X POST -d @samples/geo.json -H "content-type: application/json" 'http://localhost:12001?remove=geometry'
```

Adds in HTTP headers `Removes`, an array of remove query params and 
`Unfiltered-Content-Length`, the original content length

example:
```
Removes: [geometry]
Unfiltered-Content-Length: 3760
```

## build / run

For development, can just `go run main.go` from the `src` dir

To build a container, `docker build -t ghchinoy/filterjson .` using the 2-stage 
build Dockerfile, then to run it `docker run -d -p 8080:12001 --name filterjson ghchinoy/filterjson` (this exposes the internal `12001` 
to make the service available on `8080`)

There's also a nodejs version of the server in `nodejs` that implements only the 
`filter` param (not `remove`), and for top level JSON objects (not JSON arrays)

## web ui

There's a simple web ui in `web` that can be used to test the service. Start with 
`yarn install && yarn run start` - it looks for the service to be available at 
localhost:12001

[short video showing capabilities](https://www.useloom.com/share/4dd8b0e7a44d4589a17eb2ac6cb84580)


CHANGELOG.md
# Changelog

0.4 20180722

* adds `nodejs` dir, a nodejs express + Ramda server to do similar (JSON objects only, not arrays) as the golang server

0.3 20180721

* ui converts comma-separated filters into appropriate query params for API

0.2 20180718

* add in multi-stage dockerfile
* implement filter (vs already implemented `remove`)
* wrap handler func to indicate total time taken to process
* react ux that uses filterjson

0.1 20170717

* initial


src/main.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"time"

	"github.com/gorilla/mux"
	"github.com/rs/cors"
)

const port = 12001

func main() {
	fmt.Println("Starting filterjson service")

	r := mux.NewRouter()
	r.HandleFunc("/", timedHandler(filterHandler)).
		Methods("POST") //.
		//Queries("remove", "filter")

	fmt.Printf("Listening on port %v ...\n", port)

	//http.Handle("/", r)

	c := cors.New(cors.Options{
		AllowedOrigins: []string{
			"http://localhost:3000",
			"http://penguin.linux.test:3000",
		},
		AllowCredentials: true,
	})
	handler := c.Handler(r)

	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%v", port), handler))
}

func timedHandler(h http.HandlerFunc) http.HandlerFunc {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		defer timethis(time.Now(), fmt.Sprintf("%+v", r.URL.Query()))
		h.ServeHTTP(w, r)
	})
}

func timethis(start time.Time, name string) {
	log.Printf("%s took %s", name, time.Since(start))
}

func corsHandler(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(200)
}

func filterHandler(w http.ResponseWriter, r *http.Request) {

	// body
	data, err := ioutil.ReadAll(r.Body)
	defer r.Body.Close()
	if err != nil {
		http.Error(w, "unable to read body", http.StatusNotAcceptable)
		return
	}

	// filters
	removes := r.URL.Query()["remove"]
	filters := r.URL.Query()["filter"]

	var testpayload interface{}
	err = json.Unmarshal(data, &testpayload)
	if err != nil {
		log.Println(err)
		http.Error(w, "ugh", http.StatusInternalServerError)
		return
	}

	var payload interface{}
	switch p := testpayload.(type) {
	case map[string]interface{}: // JSON Object
		if len(filters) > 0 {
			log.Println("filtering")
			payload = returnOnlyKeysInJSON(p, filters)
			p = payload.(map[string]interface{})
		}
		if len(removes) > 0 {
			log.Println("removing")
			payload = removeKeysFromJSON(p, removes)
		}
	case []interface{}: // JSON Array
		if len(filters) > 0 {
			payload = returnOnlyKeysInJSONArray(p, filters)
			p = payload.([]interface{})
		}
		if len(removes) > 0 {
			payload = removeKeysFromJSONArray(p, removes)
		}
	default: // unknown
		log.Printf("idk wtf: %+v", p)
	}

	w.Header().Set("filters", fmt.Sprintf("%s", filters))
	w.Header().Set("removes", fmt.Sprintf("%s", removes))
	w.Header().Set("unfiltered-content-length", fmt.Sprintf("%v", len(data)))

	p, err := json.Marshal(payload)
	if err != nil {
		http.Error(w, "unable to byte removed payload", http.StatusInternalServerError)
		return
	}

	w.Write(p)
}

// removeKeysFromJSONArray Removes the list of keys in "removes" from the array
// of JSON objects
func removeKeysFromJSONArray(payload []interface{}, removes []string) []interface{} {
	var response []interface{}
	for _, v := range payload {
		r := removeKeysFromJSON(v.(map[string]interface{}), removes)
		response = append(response, r)
	}
	return response
}

// removeKeysFromJSON Removes the keys in "removes" from the JSON object given
func removeKeysFromJSON(payload map[string]interface{}, removes []string) map[string]interface{} {
	for _, r := range removes {
		delete(payload, r)
	}
	return payload
}

// returnOnlyKeysInJSON returns only the keys listed in "filters" in the given JSON object
func returnOnlyKeysInJSON(payload map[string]interface{}, filters []string) map[string]interface{} {
	kept := make(map[string]interface{})
	for _, f := range filters {
		if _, ok := payload[f]; ok == true {
			kept[f] = payload[f]
		}
	}
	return kept
}

// returnOnlyKeysInJSONArray returns an array of JSON objects with only the keys specified
// in "filters"
func returnOnlyKeysInJSONArray(payload []interface{}, filters []string) []interface{} {
	var response []interface{}
	for _, v := range payload {
		// this will error if the type conversion doesn't work
		// valid json is array of arrays, not everything is an array of objects,
		// and an array of arrays will panic here
		kept := returnOnlyKeysInJSON(v.(map[string]interface{}), filters)
		response = append(response, kept)
	}
	return response
}


web/src/pages/index.js
import React from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
//import Dialog from '@material-ui/core/Dialog';
//import DialogTitle from '@material-ui/core/DialogTitle';
//import DialogContent from '@material-ui/core/DialogContent';
//import DialogContentText from '@material-ui/core/DialogContentText';
//import DialogActions from '@material-ui/core/DialogActions';
import Typography from '@material-ui/core/Typography';
import AppBar from '@material-ui/core/AppBar'
import Toolbar from '@material-ui/core/Toolbar'
import Paper from '@material-ui/core/Paper'
import Grid from '@material-ui/core/Grid'
import TextField from '@material-ui/core/TextField';
import { withStyles } from '@material-ui/core/styles';
import withRoot from '../withRoot';

const styles = theme => ({
  root: {
    textAlign: 'center',
    //paddingTop: theme.spacing.unit * 20,
  },
  paper: {
    ...theme.mixins.gutters(),
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2,
  },
  textField: {
    marginLeft: theme.spacing.unit,
    marginRight: theme.spacing.unit,
    //width: 200,
  },
  button: {
    margin: theme.spacing.unit,
  },
});

class Index extends React.Component {
  state = {
    //open: false,
    filter: "",
    remove: "",
    unfiltered: {
      "hello": 1,
      "there": 2
    },
    filtered: {}
  }; 

  // currently only does the filters
  assembleQueryParams = () => {
    let queryParams = "?";
    let filters = this.state.filter.split(',');
    if (filters.length > 1) {
      let joined = filters.map(f => `filter=${f}`);
      queryParams += joined.join('&');
      console.log("joined", queryParams)
    } else {
      queryParams += `filter=${this.state.filter}`;
    }
    return queryParams;
  }

  handleFilterInvokation = () => {
    let service = `${window.location.protocol}//${window.location.hostname}:12001`;
    //console.log(this.state.unfiltered);
    try {
      //console.log("Attempting fetch...")
      let body = JSON.stringify(this.state.unfiltered)
      fetch(`${service}/${this.assembleQueryParams()}`,{
        method: 'post',
        headers: {'content-type': 'application/json'},
        body: body
      })
      .then( response => response.json())
      .then( data => {
        //console.log(data);
        this.setState({filtered: data});
      });
    } catch(err) {
      console.log("error",err);
    }
  };

  handleParameterChanges = key => (event, value) => {
    //console.log("paramchange", key)
    //console.log(key,event.target.value)
    this.setState({
      [key]: event.target.value,
    });
  };
  handleInputJSONChange = (e) => {
    //console.log("jsonchange")
    try {
      let unfiltered = JSON.parse(e.target.value)
      this.setState( {
        unfiltered: unfiltered
      });
    } catch(err) {
      console.log("error",err);
    }
  };

  hello = () => {
    console.log("hello");
  };

  

  render() {
    const { classes } = this.props;
    //const { open } = this.state;

    return (
      <div className={classes.root}>
        <AppBar position="static" color="default">
          <Toolbar>
            <Typography variant="title">
              FilterJSON tester
            </Typography>
          </Toolbar>
        </AppBar>
        <Paper className={classes.paper} elevation={1}>
          <Grid container spacing={24}>
            <Grid item xs={12}>
              <TextField
                label="Filter"
                id="param-filter"
                className={classes.textField}
                helperText="comma-separated keys to select"
                margin="dense"
                value={this.state.filter || ''}
                onChange={this.handleParameterChanges('filter')}
              />
              <TextField
                label="Remove"
                id="param-remove"
                className={classes.textField}
                helperText="comma-separated keys to remove"
                margin="dense"
                value={this.state.remove || ''}
                onChange={this.handleParameterChanges('remove')}
              />
              <Button 
                variant="contained" 
                color="primary" 
                onClick={this.handleFilterInvokation.bind(this)}
                className={classes.button}
                >
                Test
              </Button>
            </Grid>
            <Grid item xs={6}>
              <TextField
                id="multiline-flexible-unfiltered"
                label="Input"
                multiline
                rowsMax="20"
                fullWidth
                className={classes.textField}
                margin="normal"
                helperText="Unfiltered JSON"
                value={JSON.stringify(this.state.unfiltered, null, "  ") || ''}
                onChange={this.handleInputJSONChange}
              />
            </Grid>
            <Grid item xs={6}>
              <TextField
                id="multiline-flexible-filtered"
                label="Output"
                multiline
                rowsMax="20"
                fullWidth
                className={classes.textField}
                margin="normal"
                helperText="Filter JSON" 
                value={JSON.stringify(this.state.filtered, null, "  ") || ''}
              />
            </Grid>
          </Grid>
        </Paper>
      </div>
    );
  }
}

Index.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withRoot(withStyles(styles)(Index));
